# MACHINE LEARNING - FIRST PROJECT 

# NIKOLAOS SPYROPOULOS, AM: 3077 

In [2]:
import tensorflow as tf
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# f1_scores
f1_knn_eu = [] #knn using euclidean
f1_knn_cos = [] #knn using cosine
f1_nn = [] #for the 2 neural networks
f1_svm = [] #for the different svm 

#accuracy
acc_knn_eu = [] #knn using euclidean
acc_knn_cos = [] #knn using cosine
acc_nn = [] #for the 2 neural networks
acc_svm = [] #for the different svm 

Loading dataset

In [3]:
fashion_mnist = tf.keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()



*    The following shows there are 60,000 images , with each image represented as 28 x 28 pixels
*   Likewise, there are 60,000 labels 





In [4]:
print('The train image dataset has shape:', train_images.shape)
print('The test image dataset has shape:',test_images.shape)
print('The train label dataset has shape:', len(train_labels))
print('The test label dataset has shape:',test_labels.shape)

The train image dataset has shape: (60000, 28, 28)
The test image dataset has shape: (10000, 28, 28)
The train label dataset has shape: 60000
The test label dataset has shape: (10000,)


Classes names are not included in the dataset so we initialize them here

In [5]:
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

The **train_images** have 3 dimensions and we have to reshape them into 2d to use it in the methods knn and svm

In [6]:
nsamples, nx, ny = train_images.shape
train_images_2d = train_images.reshape((nsamples,nx*ny))
train_images_2d.shape

(60000, 784)

same as previously but for the **test_images**

In [7]:
nsamples, nx, ny = test_images.shape
test_images_2d = test_images.reshape((nsamples,nx*ny))
test_images_2d.shape

(10000, 784)

# **KNN classifier function**

In [7]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import f1_score, accuracy_score

In [8]:
def knn_classifier(k, distance):
    print('\nKNN Classifier with n_neighbors = ', k, 'using', distance, 'distance\n')
    classifier = KNeighborsClassifier(n_neighbors = k , metric = distance)
    classifier.fit(train_images_2d, train_labels)
    y_pred = classifier.predict(test_images_2d)
    
    f1_current = f1_score(test_labels,y_pred, average = 'weighted')
    acc_current = accuracy_score(test_labels,y_pred)
    if distance is 'euclidean':
        f1_knn_eu.append(f1_current)
        acc_knn_eu.append(acc_current)
    else:
        f1_knn_cos.append(f1_current)
        acc_knn_cos.append(acc_current)
            
    print("F1 Score: {}".format(f1_current))
    print("Accuracy: {}".format(acc_current))

**KNeighborsClassifier using euclidean distance for K = 1, 5, 10**

In [9]:
k_list = [1,5,10]
for k in k_list:
    knn_classifier(k, 'euclidean')


KNN Classifier with n_neighbors =  1 using euclidean distance

F1 Score: 0.8503492525016987
Accuracy: 0.8497

KNN Classifier with n_neighbors =  5 using euclidean distance

F1 Score: 0.8546439722018904
Accuracy: 0.8554

KNN Classifier with n_neighbors =  10 using euclidean distance

F1 Score: 0.8506366581732875
Accuracy: 0.8515


**KNeighborsClassifier using cosine distance for K = 1, 5, 10**

In [10]:
k_list = [1,5,10]
for k in k_list:
    knn_classifier(k, 'cosine')


KNN Classifier with n_neighbors =  1 using cosine distance

F1 Score: 0.8575689280735682
Accuracy: 0.8576

KNN Classifier with n_neighbors =  5 using cosine distance

F1 Score: 0.8559639833399547
Accuracy: 0.8578

KNN Classifier with n_neighbors =  10 using cosine distance

F1 Score: 0.8497578634159508
Accuracy: 0.8529


# KNN RESULTS

As we can see from the above results the the **F1_score with ascending order**

**EUCLIDEAN DISTANCE**
1.   **Κ = 1** F1 Score: 0.8503492525016987
2.   **K = 10** F1 Score: 0.8506366581732875
3.   **K = 5** F1 Score: 0.8546439722018904

**COSINE DISTANCE**
1.   **Κ = 10** F1 Score: 0.8497578634159508
2.   **K = 5** F1 Score: 0.8559639833399547
3.   **K = 1** F1 Score: 0.8575689280735682

So the **best case** of KNN according to the **F1_score** was by using **cosine distance with 1 neigbhor** and the **worse** was **cosine distance** with **10 neighbors** 


As we can see from the above results the the **ACCURACY with ascending order**

**EUCLIDEAN DISTANCE**
1.   **Κ = 1** Accuracy: 0.8497
2.   **K = 10** Accuracy: 0.8515
3.   **K = 5** Accuracy: 0.8554

**COSINE DISTANCE**
1.   **Κ = 10** Accuracy: 0.8529
2.   **K = 1** Accuracy: 0.8576
3.   **K = 5** Accuracy: 0.8578

So the **best case** of KNN according to the **ACCURACY** was by using **cosine distance with 5 neigbhors** and the **worse** was **euclidean distance** with **1 neighbors** 


# NEURAL NETWORKS

**Normalization** of the data to use them in our model.

In [5]:
train_images = train_images / 255.0
test_images = test_images / 255.0

**1st NEURAL NETWORK** with **sigmoid activation function** in every neuron


*   **1 Hidden** layer with **500 neurons**
*   for the training we are using Stochastic Gradient Descent
*   the **output** has **10 neurons**. The probability that a data (image) belongs to each category will be calculated by using the **softmax** activation function.







In [13]:
K = 500
first_model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(K, activation='sigmoid'),
    tf.keras.layers.Dense(10, activation='softmax')
])
opt = tf.keras.optimizers.SGD()

# first_model.compile(optimizer = opt, metrics=['accuracy'])
first_model.compile(optimizer='SGD',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
first_model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 7s 2ms/step - loss: 1.5281 - accuracy: 0.5698
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.7445 - accuracy: 0.7575
Epoch 3/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.6427 - accuracy: 0.7784
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.5814 - accuracy: 0.7960
Epoch 5/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.5564 - accuracy: 0.8044
Epoch 6/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.5245 - accuracy: 0.8198
Epoch 7/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.5179 - accuracy: 0.8194
Epoch 8/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4936 - accuracy: 0.8294
Epoch 9/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4889 - accuracy: 0.8314
Epoch 10/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.483

In [14]:
test_loss, test_acc = first_model.evaluate(test_images,  test_labels, verbose=2)
print('\nTest accuracy:', test_acc)
print('\nTest loss:', test_loss)

313/313 - 3s - loss: 0.5015 - accuracy: 0.8224

Test accuracy: 0.8223999738693237

Test loss: 0.5014984011650085


In [15]:
from sklearn.metrics import f1_score, accuracy_score
y_pred = first_model.predict(test_images, batch_size=64, verbose=1)
y_pred_bool = np.argmax(y_pred, axis=1)

f1_current =f1_score(test_labels,y_pred_bool, average = 'weighted')
f1_nn.append(f1_current)

acc_current = accuracy_score(test_labels,y_pred_bool)
acc_nn.append(acc_current)

print("F1 Score of the 1st model: {}".format(f1_current))
print("Accuracy of the 1st model: {}".format(acc_current))


157/157 [==============================] - 3s 2ms/step
F1 Score of the 1st model: 0.8195194562895265
Accuracy of the 1st model: 0.8224


**2nd NEURAL NETWORK** with **sigmoid activation function** in every neuron


*   **1 Hidden** layer with **500 neurons**
*   **2 Hidden** layer with **200 neurons**
*   for the training we are using Stochastic Gradient Descent
*   the **output** has **10 neurons**. The probability that a data (image) belongs to each category will be calculated by using the **softmax** activation function.







In [16]:
K1 = 500
K2 = 200
second_model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(K1, activation='sigmoid'),
    tf.keras.layers.Dense(K2, activation='sigmoid'),
    tf.keras.layers.Dense(10, activation='softmax')
])
opt = tf.keras.optimizers.SGD()

second_model.compile(optimizer='SGD',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
second_model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 7s 2ms/step - loss: 2.0840 - accuracy: 0.3542
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 1.1892 - accuracy: 0.6463
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.9007 - accuracy: 0.7104
Epoch 4/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.7580 - accuracy: 0.7408
Epoch 5/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6916 - accuracy: 0.7509
Epoch 6/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.6439 - accuracy: 0.7671
Epoch 7/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6196 - accuracy: 0.7749
Epoch 8/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5851 - accuracy: 0.7886
Epoch 9/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5684 - accuracy: 0.7963
Epoch 10/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.543

In [17]:
test_loss, test_acc = second_model.evaluate(test_images,  test_labels, verbose=2)
print('\nTest accuracy:', test_acc)

313/313 - 3s - loss: 0.5566 - accuracy: 0.7997

Test accuracy: 0.7997000217437744


In [18]:
from sklearn.metrics import f1_score, accuracy_score
y_pred = second_model.predict(test_images, batch_size=64, verbose=1)
y_pred_bool = np.argmax(y_pred, axis=1)

f1_current =f1_score(test_labels,y_pred_bool, average = 'weighted')
f1_nn.append(f1_current)

acc_current = accuracy_score(test_labels,y_pred_bool)
acc_nn.append(acc_current)

print("F1 Score of the 2nd model: {}".format(f1_current))
print("Accuracy of the 2nd model: {}".format(acc_current))

157/157 [==============================] - 3s 2ms/step
F1 Score of the 2nd model: 0.795980740331989
Accuracy of the 2nd model: 0.7997


# NEURAL NETWORKS RESULTS

As we can see from the above results:

the **F1_score with ascending order**

1.   **Second model** F1 Score: 0.795980740331989
2.   **First model** F1 Score of the 1st model: 0.8195194562895265


the **ACCURACY with ascending order**
1.   **Second model** Accuracy of the 2nd model: 0.7997
2.   **First model** Accuracy of the 1st model: 0.8224


So the **best case** of **neural networks** according to the **F1_score** and the **Accuracy** is the **First_model** 

# SVM

**LINEAR KERNEL**

In [16]:
from sklearn.svm import SVC
# svc_linear_kernel = LinearSVC()

svc_linear_kernel = SVC(kernel='linear', C = 1.0)
svc_linear_kernel.fit(train_images_2d, train_labels)

SVC(kernel='linear')

**Making prediction**

In [23]:
y_pred = svc_linear_kernel.predict(test_images_2d)

**Calculate metrics**

In [25]:
from sklearn.metrics import f1_score, accuracy_score
print("F1 Score: {}".format(f1_score(test_labels,y_pred, average = 'weighted')))
print("Accuracy: {}".format(accuracy_score(test_labels,y_pred)))

F1 Score: 0.8456969698281024
Accuracy: 0.8464


**GAUSSIAN KERNEL**

In [8]:
from sklearn.svm import SVC
svc_gaussian_kernel = SVC(kernel='rbf')
svc_gaussian_kernel.fit(train_images_2d, train_labels)

SVC()

In [10]:
y_pred = svc_gaussian_kernel.predict(test_images_2d)

In [11]:
from sklearn.metrics import f1_score, accuracy_score
print("F1 Score with gaussian kernel: {}".format(f1_score(test_labels,y_pred, average = 'weighted')))
print("Accuracy with gaussian kernel: {}".format(accuracy_score(test_labels,y_pred)))

F1 Score with gaussian kernel: 0.8822648793630384
Accuracy with gaussian kernel: 0.8828


**COSINE KERNEL** 

In [8]:
from sklearn.svm import SVC
from sklearn.metrics.pairwise import cosine_similarity
# cosine_kernel = cosine_similarity(train_images, train_labels)
svc_cosine_kernel = SVC(kernel= cosine_similarity)
# svc_cosine_kernel.fit(train_images, train_labels)

In [9]:
svc_cosine_kernel.fit(train_images_2d[:30000], train_labels[:30000])

SVC(kernel=<function cosine_similarity at 0x00000189D1FA3288>)

In [ ]:
y_pred = svc_cosine_kernel.predict(train_images_2d[:30000])

# SVM RESULTS

As we can see from the above results:

the **F1_score with ascending order**

1.   **LINEAR KERNEL** F1 Score: 0.8456969698281024
2.   **GAUSSIAN KERNEL** F1 Score with gaussian kernel: 0.8822648793630384


the **ACCURACY with ascending order**
1.   **LINEAR KERNEL** Accuracy: 0.8464
2.   **GAUSSIAN KERNEL** Accuracy with gaussian kernel: 0.8828


So the **best case** of **SVM** according to the **F1_score** and the **Accuracy** is the **GAUSSIAN KERNEL** 

# FINAL RESULTS

After examining all our methods we came to the following conclusion:

***we are going to compare the max metrics that we found previously in KNN and neural networks with the metric of the svm***

the **F1_score with ascending order**
***The best of the 3 methods are***

1.   **First model** F1 Score of the 1st model: 0.8195194562895265
2.   **knn-cosine distance with 1 neigbhor** F1 Score: 0.8575689280735682
3.   **svm-gaussian kernel**  F1 Score with gaussian kernel: 0.8822648793630384

the **ACCURACY with ascending order**
1.   **First model** Accuracy of the 1st model: 0.8224
2.   **knn-cosine distance with 5 neigbhors** Accuracy: 0.8578
3.   **svm-gaussian-kernel** Accuracy with gaussian kernel: 0.8828

So the **best case** according to the methods that we used that has the best **F1_score** and the **Accuracy** is **svm-gaussian kernel** 